metrics/internal/stackdriver/secret_watcher_test.go

```diff
@@ -0,0 +1,569 @@
+	defaultNumTestObservers = 2
+
+	defaultGlobalSecretWatcherConstructor = func(t *testing.T, obs ...Observer) SecretWatcher {
+		return mustNewSecretWatcher(t, obs...)
+	}
+
+	defaultSingleNamespaceSecretWatcherConstructor = func(t *testing.T, obs ...Observer) SecretWatcher {
+		return mustNewSecretWatcherSingleNamespace(t, dsft.namespace, obs...)
+	}
+
+	defaultSingleSecretSecretWatcherConstructor = func(t *testing.T, obs ...Observer) SecretWatcher {
+		return mustNewSecretWatcherSingleSecret(t, dsft.namespace, dsft.name, obs...)
+	}
+
+	defaultWatchersToTest = []struct {
+		name        string
+		constructor func(t *testing.T, obs ...Observer) SecretWatcher
+	}{
+		{
+			name:        "GlobalSecretWatcher",
+			constructor: defaultGlobalSecretWatcherConstructor,
+		}, {
+			name:        "SingleNamespaceSecretWatcher",
+			constructor: defaultSingleSecretSecretWatcherConstructor,
+		}, {
+			name:        "SingleSecretSecretWatcher",
+			constructor: defaultSingleSecretSecretWatcherConstructor,
+		},
+	}
+
+	defaultObserverFuncs = &ObserverFuncs{
+		AddFunc: func(s *corev1.Secret) {
+		},
+		UpdateFunc: func(sOld *corev1.Secret, sNew *corev1.Secret) {
+		},
+		DeleteFunc: func(s *corev1.Secret) {
+		},
+	}
+)
+
+// secretForTest encapsulates Secret metadata and a Secret for re-using and verifying values.
+type secretForTest struct {
+	namespace string
+	name      string
+	dataKey   string
+	dataValue string
+	secret    corev1.Secret
+}
+
+func (s *secretForTest) UpdateDataKey(dataKey string) *secretForTest {
+	s.dataKey = dataKey
+	s.secret.Data = map[string][]byte{
+		s.dataKey: []byte(s.dataValue),
+	}
+
+	return s
+}
+
+func (s *secretForTest) UpdateDataValue(dataValue string) *secretForTest {
+	s.dataValue = dataValue
+	s.secret.Data = map[string][]byte{
+		s.dataKey: []byte(s.dataValue),
+	}
+
+	return s
+}
+
+// newSecretForTest constructs a secretForTest.
+func newSecretForTest(namespace string, name string) *secretForTest {
+	return &secretForTest{
+		namespace: namespace,
+		name:      name,
+		dataKey:   defaultSecretDataKey,
+		dataValue: defaultSecretDataValue,
+		secret: corev1.Secret{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Secret",
+				APIVersion: "apps/v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+			Data: map[string][]byte{
+				defaultSecretDataKey: []byte(defaultSecretDataValue),
+			},
+			Type: "Opaque",
+		},
+	}
+}
+
+// testObserver is an implementation of the the secretWatcher observer interface
+// that also holds state about whether the observer's callbacks have been triggered.
+type testObserver struct {
+	oFuncs         *ObserverFuncs
+	onAddCalled    bool
+	onUpdateCalled bool
+	onDeleteCalled bool
+	mux            sync.Mutex
+}
+
+// newTestObserver constructs a testObserver from observerFuncs.
+func newTestObserver(inputObsFuncs *ObserverFuncs) *testObserver {
+	// If a callback is nil, consider it to be called by default.
```

Comment:
Why default to true when a callback is nil?

Reply:
It is for the conveinence of the AllCallbacksCalled() helper function. If only the AddFunc is defined, I want that to still return "true".

Reply:
Turns out I stopped using AllCallbacksCalled() at some point, so I've removed that function all together and removed the "true" default for nil funcs.